File: DTCore5.0-exp/DTCore/DataTools.Memory/StringBlob/BSTR.vb
'' ************************************************* ''
'' DataTools Visual Basic Utility Library 
''
'' Module: BSTR
''         Length-preambled text string.
'' 
'' Copyright (C) 2011-2020 Nathan Moschkin
'' All Rights Reserved
''
'' Licensed Under the Microsoft Public License   
'' ************************************************* ''

Imports DataTools.Memory.Internal
Imports System.Runtime.InteropServices

Namespace Memory


#Region "BSTR"

    ''' <summary>
    ''' BSTR implementation
    ''' </summary>
    ''' <remarks></remarks>
    Public Class BSTR
        Implements IDisposable

        Friend _ptr As IntPtr
        Friend _preamble As Integer = 4
        Friend _Com As Boolean = False

        Public Overrides Function ToString() As String
            Return _getString()
        End Function

        Public Sub New(ptr As IntPtr)
            _ptr = ptr
        End Sub

        Public Sub New(ptr As IntPtr, fromCom As Boolean)
            _ptr = ptr
            If (fromCom) Then
                _Com = True
                _ptr -= 4
                _preamble = 4
            End If
        End Sub

        Public Sub New()

        End Sub

        Public Property PreambleLength As Integer
            Get
                Return _preamble
            End Get
            Set(value As Integer)
                If _Com Then
                    _preamble = 4
                    Return
                End If

                If (Length <> 0) AndAlso (value < _preamble) AndAlso (Length > _preMax(value)) Then
                    Throw New ArgumentException("Cannot downgrade preamble with long string present.")
                End If

                If _ptr = IntPtr.Zero Then
                    _preamble = value
                    Return
                End If

                Dim tl As Integer = Length
                Dim l As Integer = tl + value
                Dim s As String = Me,
                    pNew As IntPtr

                pNew = HeapAlloc(GetProcessHeap, 0, l)

                CopyMemory(pNew, tl, CIntPtr(value))
                CopyMemory(IntPtr.Add(pNew, value), IntPtr.Add(_ptr, CIntPtr(_preamble)), CIntPtr(tl * 2))
                HeapFree(GetProcessHeap, 0, _ptr)
                _ptr = pNew
                _preamble = value
            End Set
        End Property

        Public Property Com As Boolean
            Get
                Return _Com
            End Get
            Friend Set(value As Boolean)
                If _Com = True Then Return

                _Com = value
                If value Then
                    _preamble = 4
                    If _ptr <> IntPtr.Zero Then
                        _ptr -= 4
                    End If
                Else
                    If _ptr <> IntPtr.Zero Then
                        _ptr += 4
                    End If
                End If
            End Set
        End Property

        Private Function _preMax(val As Integer) As Integer

            Select Case val

                Case 1
                    Return 255

                Case 2
                    Return 32767

                Case 4
                    Return &H7FFFFFFF

                Case Else
                    Return -1L

            End Select

        End Function

        Public ReadOnly Property Handle As IntPtr
            Get
                If (_Com) Then Handle = _ptr + 4 Else Handle = _ptr
            End Get
        End Property

        Public ReadOnly Property Length As Integer
            Get
                If _ptr = IntPtr.Zero Then Return -1
                CopyMemory(Length, _ptr, CIntPtr(_preamble))
            End Get
        End Property

        Public Property Text As String
            Get
                Text = _getString()
            End Get
            Set(value As String)
                _setString(value)
            End Set
        End Property

        Private Sub _setString(value As String)
            If value.Length > _preMax(_preamble) Then
                Throw New ArgumentException("Long string won't fit current preamble.")
            End If

            Dim nl As Integer = (value.Length * 2) + _preamble
            Dim hHeap As IntPtr = GetProcessHeap()

            '' already have a string, let's see what is going on...
            If _ptr <> IntPtr.Zero Then
                '' get the heap buffer size.
                Dim lb As Integer = HeapSize(hHeap, 0, _ptr)

                If value = "" Then
                    ' null string, release all memory and tell the garbage collector
                    HeapFree(hHeap, 0, _ptr)
                    GC.RemoveMemoryPressure(lb)

                    Return
                End If

                '' we're actually reallocating the buffer.
                If lb <> nl Then
                    _ptr = HeapReAlloc(hHeap, 0, _ptr, nl)


                    '' have to check the actual allocated size because it may be different from the calculated length
                    nl = HeapSize(hHeap, 0, _ptr)
                    If nl > lb Then
                        '' tell the garbage collector about the change in size
                        GC.AddMemoryPressure(nl - lb)
                    Else
                        '' tell the garbage collector about the change in size
                        GC.RemoveMemoryPressure(lb - nl)
                    End If
                End If
            Else
                '' allocate a new buffer!
                _ptr = HeapAlloc(hHeap, 0, nl)

                '' have to check the actual allocated size because it may be different from the calculated length
                nl = HeapSize(hHeap, 0, _ptr)

                '' tell the garbage collector about the allocation
                GC.AddMemoryPressure(nl)
            End If

            Dim l As Integer = value.Length
            Dim p As IntPtr = IntPtr.Add(_ptr, CIntPtr(_preamble))

            '' copy the length of the string (in characters) into the preamble
            CopyMemory(_ptr, l, CIntPtr(_preamble))

            '' copy the string into the buffer after the preamble.
            CopyMemory(p, value, CIntPtr(value.Length * 2))
        End Sub

        Private Function _getString() As String
            Dim l As Short
            Dim s As String

            '' get the length of the string out of the preamble.
            CopyMemory(l, _ptr, CIntPtr(_preamble))

            s = New String(ChrW(0), l)
            Dim ptr As IntPtr = IntPtr.Add(_ptr, CIntPtr(_preamble))

            CopyMemory(s, ptr, CIntPtr(l * 2))
            Return s

        End Function

        Public Shared Narrowing Operator CType(operand As BSTR) As IntPtr
            Return operand.Handle
        End Operator

        Public Shared Narrowing Operator CType(operand As IntPtr) As BSTR
            Return New BSTR(operand, True)
        End Operator

        Public Shared Narrowing Operator CType(operand As BSTR) As MemPtr
            Return operand.Handle
        End Operator

        Public Shared Narrowing Operator CType(operand As MemPtr) As BSTR
            Return New BSTR(operand, True)
        End Operator

        Public Shared Narrowing Operator CType(operand As BSTR) As UIntPtr
            Return New UIntPtr(CULng(operand.Handle))
        End Operator

        Public Shared Narrowing Operator CType(operand As UIntPtr) As BSTR
            Return New BSTR(New IntPtr(CType(operand.ToUInt64, Long)), True)
        End Operator

        Public Shared Narrowing Operator CType(operand As String) As BSTR
            Dim b As New BSTR
            b._setString(operand)
            Return b
        End Operator

        Public Shared Narrowing Operator CType(operand As Char()) As BSTR
            Dim b As New BSTR
            b._setString(CStr(operand))
            Return b
        End Operator

        Public Shared Narrowing Operator CType(operand As BSTR) As String
            Return operand._getString()
        End Operator

        Public Shared Narrowing Operator CType(operand As BSTR) As Char()
            Dim l As Short
            Dim ch() As Char

            '' get the size from the preamble
            CopyMemory(l, operand._ptr, CIntPtr(2))

            ReDim ch(l - 1)

            '' get the pointer to memory just past the preamble
            Dim ptr As IntPtr = IntPtr.Add(operand._ptr, CIntPtr(2))

            '' copy the memory into the managed array.
            CopyMemory(ch, ptr, CIntPtr(l * 2))

            '' return the new array
            Return ch
        End Operator

        '' same as above but with bytes.
        Public Shared Narrowing Operator CType(operand As BSTR) As Byte()
            Dim i As Integer = operand.Length,
                b() As Byte
            Dim l As Integer = (i * 2) + operand._preamble

            ReDim b(l - 1)
            CopyMemory(b, operand._ptr, CIntPtr(l))
            Return b
        End Operator

        Public Shared Narrowing Operator CType(operand As Byte()) As BSTR
            Dim bs As New BSTR
            Dim hHeap As IntPtr = GetProcessHeap()

            bs._ptr = HeapAlloc(hHeap, 0, operand.Length)
            CopyMemory(bs._ptr, operand.Length, CIntPtr(bs._preamble))
            CopyMemory(IntPtr.Add(bs._ptr, bs._preamble), operand, CIntPtr(operand.Length))

            '' get the size of the allocated buffer and tell the garbage collector
            Dim bl As Long = HeapSize(hHeap, 0, bs._ptr)
            GC.AddMemoryPressure(bl)

            Return bs
        End Operator

#Region "IDisposable Support"
        Private disposedValue As Boolean ' To detect redundant calls

        ' IDisposable
        Protected Overridable Sub Dispose(disposing As Boolean)
            If Not Me.disposedValue Then

                If _ptr <> IntPtr.Zero Then

                    If _Com Then
                        Marshal.FreeCoTaskMem(_ptr + 4)
                    Else
                        Dim hHeap As IntPtr = GetProcessHeap()
                        '' free up the unmanaged buffer 
                        Dim l As Long = HeapSize(hHeap, 0, _ptr)

                        If HeapFree(hHeap, 0, _ptr) = IntPtr.Zero Then
                            GC.RemoveMemoryPressure(l)
                        End If

                    End If

                End If
            End If

            Me.disposedValue = True
        End Sub

        Protected Overrides Sub Finalize()
            ' Do not change this code.  Put cleanup code in Dispose(disposing As Boolean) above.
            Dispose(False)
            MyBase.Finalize()
        End Sub

        ' This code added by Visual Basic to correctly implement the disposable pattern.
        Public Sub Dispose() Implements IDisposable.Dispose
            ' Do not change this code.  Put cleanup code in Dispose(disposing As Boolean) above.
            Dispose(True)
            GC.SuppressFinalize(Me)
        End Sub
#End Region

    End Class

#End Region
End Namespace

File: DTCore5.0-exp/DTCore/DataTools.Memory/StringBlob/BSTR.cs
// ' ************************************************* ''
// ' DataTools Visual Basic Utility Library 
// '
// ' Module: BSTR
// '         Length-preambled text string.
// ' 
// ' Copyright (C) 2011-2020 Nathan Moschkin
// ' All Rights Reserved
// '
// ' Licensed Under the Microsoft Public License   
// ' ************************************************* ''

using System;
using System.Runtime.InteropServices;
using DataTools.Memory.Internal;
using Microsoft.VisualBasic.CompilerServices;

namespace DataTools.Memory
{


    /* TODO ERROR: Skipped RegionDirectiveTrivia */
    /// <summary>
    /// BSTR implementation
    /// </summary>
    /// <remarks></remarks>
    public class BSTR : IDisposable
    {
        internal IntPtr _ptr;
        internal int _preamble = 4;
        internal bool _Com = false;

        public override string ToString()
        {
            return _getString();
        }

        public BSTR(IntPtr ptr)
        {
            _ptr = ptr;
        }

        public BSTR(IntPtr ptr, bool fromCom)
        {
            _ptr = ptr;
            if (fromCom)
            {
                _Com = true;
                _ptr = _ptr - 4;
                _preamble = 4;
            }
        }

        public BSTR()
        {
        }

        public int PreambleLength
        {
            get
            {
                return _preamble;
            }

            set
            {
                if (_Com)
                {
                    _preamble = 4;
                    return;
                }

                if (Length != 0 && value < _preamble && Length > _preMax(value))
                {
                    throw new ArgumentException("Cannot downgrade preamble with long string present.");
                }

                if (_ptr == IntPtr.Zero)
                {
                    _preamble = value;
                    return;
                }

                int tl = Length;
                int l = tl + value;
                string s = (string)this;
                IntPtr pNew;
                pNew = Native.HeapAlloc(Native.GetProcessHeap(), 0U, (IntPtr)l);
                Native.CopyMemory(pNew, ref tl, Native.CIntPtr(value));
                Native.CopyMemory(IntPtr.Add(pNew, value), IntPtr.Add(_ptr, (int)Native.CIntPtr(_preamble)), Native.CIntPtr(tl * 2));
                Native.HeapFree(Native.GetProcessHeap(), 0U, _ptr);
                _ptr = pNew;
                _preamble = value;
            }
        }

        public bool Com
        {
            get
            {
                return _Com;
            }

            internal set
            {
                if (_Com == true)
                    return;
                _Com = value;
                if (value)
                {
                    _preamble = 4;
                    if (_ptr != IntPtr.Zero)
                    {
                        _ptr = _ptr - 4;
                    }
                }
                else if (_ptr != IntPtr.Zero)
                {
                    _ptr = _ptr + 4;
                }
            }
        }

        private int _preMax(int val)
        {
            switch (val)
            {
                case 1:
                    {
                        return 255;
                    }

                case 2:
                    {
                        return 32767;
                    }

                case 4:
                    {
                        return 0x7FFFFFFF;
                    }

                default:
                    {
                        return (int)-1L;
                    }
            }
        }

        public IntPtr Handle
        {
            get
            {
                IntPtr HandleRet = default;
                if (_Com)
                    HandleRet = _ptr + 4;
                else
                    HandleRet = _ptr;
                return HandleRet;
            }
        }

        public int Length
        {
            get
            {
                int LengthRet = default;
                if (_ptr == IntPtr.Zero)
                    return -1;
                Native.CopyMemory(ref LengthRet, _ptr, Native.CIntPtr(_preamble));
                return LengthRet;
            }
        }

        public string Text
        {
            get
            {
                string TextRet = default;
                TextRet = _getString();
                return TextRet;
            }

            set
            {
                _setString(value);
            }
        }

        private void _setString(string value)
        {
            if (value.Length > _preMax(_preamble))
            {
                throw new ArgumentException("Long string won't fit current preamble.");
            }

            int nl = value.Length * 2 + _preamble;
            var hHeap = Native.GetProcessHeap();

            // ' already have a string, let's see what is going on...
            if (_ptr != IntPtr.Zero)
            {
                // ' get the heap buffer size.
                int lb = (int)Native.HeapSize(hHeap, 0U, _ptr);
                if (string.IsNullOrEmpty(value))
                {
                    // null string, release all memory and tell the garbage collector
                    Native.HeapFree(hHeap, 0U, _ptr);
                    GC.RemoveMemoryPressure(lb);
                    return;
                }

                // ' we're actually reallocating the buffer.
                if (lb != nl)
                {
                    _ptr = Native.HeapReAlloc(hHeap, 0, _ptr, (IntPtr)nl);


                    // ' have to check the actual allocated size because it may be different from the calculated length
                    nl = (int)Native.HeapSize(hHeap, 0U, _ptr);
                    if (nl > lb)
                    {
                        // ' tell the garbage collector about the change in size
                        GC.AddMemoryPressure(nl - lb);
                    }
                    else
                    {
                        // ' tell the garbage collector about the change in size
                        GC.RemoveMemoryPressure(lb - nl);
                    }
                }
            }
            else
            {
                // ' allocate a new buffer!
                _ptr = Native.HeapAlloc(hHeap, 0U, (IntPtr)nl);

                // ' have to check the actual allocated size because it may be different from the calculated length
                nl = (int)Native.HeapSize(hHeap, 0U, _ptr);

                // ' tell the garbage collector about the allocation
                GC.AddMemoryPressure(nl);
            }

            int l = value.Length;
            var p = IntPtr.Add(_ptr, (int)Native.CIntPtr(_preamble));

            // ' copy the length of the string (in characters) into the preamble
            Native.CopyMemory(_ptr, ref l, Native.CIntPtr(_preamble));

            // ' copy the string into the buffer after the preamble.
            Native.CopyMemory(p, value, Native.CIntPtr(value.Length * 2));
        }

        private string _getString()
        {
            var l = default(short);
            string s;

            // ' get the length of the string out of the preamble.
            Native.CopyMemory(ref l, _ptr, Native.CIntPtr(_preamble));
            s = new string('\0', l);
            var ptr = IntPtr.Add(_ptr, (int)Native.CIntPtr(_preamble));
            Native.CopyMemory(s, ptr, Native.CIntPtr(l * 2));
            return s;
        }

        public static explicit operator IntPtr(BSTR operand)
        {
            return operand.Handle;
        }

        public static explicit operator BSTR(IntPtr operand)
        {
            return new BSTR(operand, true);
        }

        public static explicit operator MemPtr(BSTR operand)
        {
            return operand.Handle;
        }

        public static explicit operator BSTR(MemPtr operand)
        {
            return new BSTR(operand, true);
        }

        public static explicit operator UIntPtr(BSTR operand)
        {
            return new UIntPtr((ulong)operand.Handle);
        }

        public static explicit operator BSTR(UIntPtr operand)
        {
            return new BSTR(new IntPtr(Conversions.ToLong(operand.ToUInt64())), true);
        }

        public static explicit operator BSTR(string operand)
        {
            var b = new BSTR();
            b._setString(operand);
            return b;
        }

        public static explicit operator BSTR(char[] operand)
        {
            var b = new BSTR();
            b._setString(Conversions.ToString(operand));
            return b;
        }

        public static explicit operator string(BSTR operand)
        {
            return operand._getString();
        }

        public static explicit operator char[](BSTR operand)
        {
            var l = default(short);
            char[] ch;

            // ' get the size from the preamble
            Native.CopyMemory(ref l, operand._ptr, Native.CIntPtr(2));
            ch = new char[l];

            // ' get the pointer to memory just past the preamble
            var ptr = IntPtr.Add(operand._ptr, (int)Native.CIntPtr(2));

            // ' copy the memory into the managed array.
            Native.CopyMemory(ch, ptr, Native.CIntPtr(l * 2));

            // ' return the new array
            return ch;
        }

        // ' same as above but with bytes.
        public static explicit operator byte[](BSTR operand)
        {
            int i = operand.Length;
            byte[] b;
            int l = i * 2 + operand._preamble;
            b = new byte[l];
            Native.CopyMemory(b, operand._ptr, Native.CIntPtr(l));
            return b;
        }

        public static explicit operator BSTR(byte[] operand)
        {
            var bs = new BSTR();
            var hHeap = Native.GetProcessHeap();
            bs._ptr = Native.HeapAlloc(hHeap, 0U, (IntPtr)operand.Length);
            int argpSrc = operand.Length;
            Native.CopyMemory(bs._ptr, ref argpSrc, Native.CIntPtr(bs._preamble));
            operand.Length = argpSrc;
            Native.CopyMemory(IntPtr.Add(bs._ptr, bs._preamble), operand, Native.CIntPtr(operand.Length));

            // ' get the size of the allocated buffer and tell the garbage collector
            long bl = (long)Native.HeapSize(hHeap, 0U, bs._ptr);
            GC.AddMemoryPressure(bl);
            return bs;
        }

        /* TODO ERROR: Skipped RegionDirectiveTrivia */
        private bool disposedValue; // To detect redundant calls

        // IDisposable
        protected virtual void Dispose(bool disposing)
        {
            if (!disposedValue)
            {
                if (_ptr != IntPtr.Zero)
                {
                    if (_Com)
                    {
                        Marshal.FreeCoTaskMem(_ptr + 4);
                    }
                    else
                    {
                        var hHeap = Native.GetProcessHeap();
                        // ' free up the unmanaged buffer 
                        long l = (long)Native.HeapSize(hHeap, 0U, _ptr);
                        if ((IntPtr)Native.HeapFree(hHeap, 0U, _ptr) == IntPtr.Zero)
                        {
                            GC.RemoveMemoryPressure(l);
                        }
                    }
                }
            }

            disposedValue = true;
        }

        ~BSTR()
        {
            // Do not change this code.  Put cleanup code in Dispose(disposing As Boolean) above.
            Dispose(false);
        }

        // This code added by Visual Basic to correctly implement the disposable pattern.
        public void Dispose()
        {
            // Do not change this code.  Put cleanup code in Dispose(disposing As Boolean) above.
            Dispose(true);
            GC.SuppressFinalize(this);
        }
        /* TODO ERROR: Skipped EndRegionDirectiveTrivia */
    }

    /* TODO ERROR: Skipped EndRegionDirectiveTrivia */
}